keps/sig-storage/5040-remove-gitrepo-driver/README.md

```diff
@@ -0,0 +1,1080 @@
+know that this has succeeded?
+-->
+
+- Remove in-tree gitRepo volume support.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Remove gitRepo volumes from the Kubernetes API.
+
+## Proposal
+
+Kubernetes drops support for gitRepo volume types by removing the in-tree driver
+that actually mounts the volume in the pod. In this case we will still keep the
+gitRepo volume in the API. Pods with gitRepo volumes will be admitted by 
+kube-apiserver however, kubelet will fail to run them with an appropriate 
+error message.
+
+
+This removal is similar to removal of support for
+[glusterfs](https://kubernetes.io/docs/concepts/storage/volumes/#glusterfs)
+which was deprecated in v1.25 and removed in v1.26.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The biggest risk is in breaking users. Despite being deprecated for years, there
+may be some users who are still using it and their workloads would be broken 
+when their clusters are upgraded. 
+
+We plan to mitigate the risk by:
+1. Allowing users to opt-in to re-enabling the driver for 3 versions to give
+  them enough time to fix  workloads
+2. Announcing the change via release notes, updates to the gitRepo documentation
+  and publishing kubernetes release blog for this feature so that customers are
+  aware and can take action before they upgrade
+3. We will update the gitRepo volume documentation to say it say that the driver
+  has been disabled by default and add information about alternative approaches.
+  We will also document how users can turn on the gitRepo volume driver however,
+  we will add a warning educating users on why they shouldn't do that.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We will add a new feature-gate to kubelet called `GitRepoVolumeDriver`. This
+feature-gate will be defaulted to `false`, which will disable the 
+[gitRepo driver][gitRepo driver](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/git_repo/git_repo.go).
+If a workload with gitRepo volume is encountered, kubelet will fail to run such a
+pod with an appropriate error message guiding users to either use an alternative
+or setting the feature-gate `GitRepoVolumeDriver` to `true`.
+
+### Validating Admission Policy
+
+Users can prevent workloads with gitRepo volumes from being admitted 
```

Comment:
Should we discuss creating a git repo to hold VAPs like this?

Reply:
Should we just put the VAP in the gitRepo volume documentation [here](https://kubernetes.io/docs/concepts/storage/volumes/#gitrepo)?

Reply:
@thockin I think we should, and we should try to make that allow for a future, more rich way to manage add ons.

I know maybe WAGNI (we ain't gonna need it), but at least some demand _is_ there.

> Should we just put the VAP in the gitRepo volume documentation [here](https://kubernetes.io/docs/concepts/storage/volumes/#gitrepo)?

No, partly because of https://github.com/kubernetes/website/issues/42590 (the detailed list of volume types doesn't really belong in that page) and partly because I wouldn't put it there.
We could put the VAP in the removal announcement and make it evergreen; after we do that, other pages can hyperlink there. Hyperlinking to more detail is the right approach.

Reply:
> Should we discuss creating a git repo to hold VAPs like this?

Yes, and we should narrowly scope it to avoid it becoming bikeshed bait, sigs.k8s.io/vaps or something like that.

> @thockin I think we should, and we should try to make that allow for a future, more rich way to manage add ons.

Addon-management is a bit of a contentious quagmire .... I think we should steer clear of that for now and just provide official reference VAPs, alongside some documentation explaining the purpose and `kubectl apply` (or bring your own installer, out of scope).

Reply:
@sftim pointed out that we may also want mutating admission policies later, `sigs.k8s.io/vaps` is probably not the right name, maybe `sigs.k8s.io/admission-policies` or something.

We should raise a repo creation request with the relevant SIG(s)?

Reply:
Which SIG(s)?  These could be any or all.  I was going to shoot a note to sig-arch this week.

Reply:
Definitely arch, perhaps security given at least the first one we'd propose there? 
I imagine cluster lifecycle would appreciate at least FYI as well, though that may steer things into the addon-management topic :+)

Reply:
[Also, this is not a blocker for the KEP, I think it's a good idea to publish a VAP, and we need to find an appropriate home, but I don't think the repo needs to be a hard requirement]

Reply:
I'd want to block off-by-default on having this in place. But not a merge of this PR.

Reply:
Yeah, I think we should plan to publish a VAP somewhere, and that should exist by off-by-default, but for alpha it can be a detail we're working out.

Reply:
Added that we plan to publish the VAPs to a public repository both in the VAP section and in the graduation criteria section.

Reply:
I don’t think this is something to discuss here, but I was a bit curious.

Kubernetes's CEL functionality is enhanced with each version, and [some features and functions become available only in certain Kubernetes versions or later](https://kubernetes.io/docs/reference/using-api/cel/#cel-options-language-features-and-libraries). If VAPs were to be published in a public repository, it would likely be necessary to explicitly state which Kubernetes version it supports.

This wouldn’t be an issue if newer CEL features weren’t used in VAPs, but for cases like the feature proposed in https://github.com/kubernetes/kubernetes/issues/129939, it would become relevant.

Reply:
Yeah, this came up in another context last week - we may need CEL-expressions to come with minimum version requirements.